src/compile/inference/imports.ts

```diff
@@ -38,66 +33,97 @@ function isPathWithRelativePrefix(path: string): boolean {
+
+    while (
+        strippedSegments < importedSegments.length &&
+        importedSegments[strippedSegments] === ".."
+    ) {
+        prefix = dirname(prefix);
+        strippedSegments++;
+    }
+
+    // According to https://docs.soliditylang.org/en/v0.8.8/path-resolution.html#relative-imports when prefix
+    // is empty there is no leading "./". However `dirname` always returns non-empty prefixes.
+    // Handle this edge case.
+    if (prefix === "." && !importer.startsWith(".")) {
+        prefix = "";
+    }
+
+    assert(prefix === "" || !prefix.endsWith("/"), `Invalid prefix ${prefix}`);
+    const suffix = importedSegments.slice(strippedSegments).join("/");
+    return prefix === "" ? suffix : prefix + "/" + suffix;
+}
+
+/**
+ * Given the `importer` source unit path, and the path `imported` of an import
+ * directive compute the expected **source unit name** of the imported file.
+ * This is the name the compiler will look for in its "VFS" as defined starting
+ * here:
+ *
+ * https://docs.soliditylang.org/en/v0.8.8/path-resolution.html
  *
- * This takes into account remappings, relative and absolute paths.
- * Note: The returned path is not neccessarily absolute!
+ * This takes into account relative and absolute paths and remappings.
  */
-function computeRealPath(
-    importer: string | undefined,
+function computeSourceUnitName(
+    importerSourceUnit: string,
     imported: string,
     remappings: Remapping[]
 ): string {
-    let result = applyRemappings(remappings, imported);
-
-    if (importer !== undefined && isPathWithRelativePrefix(result)) {
-        const importingFileDir = dirname(importer);
-
-        result = normalizeImportPath(join(importingFileDir, result));
+    if (isPathWithRelativePrefix(imported)) {
+        return normalizeRelativeImportPath(importerSourceUnit, imported);
     }
 
-    return result;
+    return applyRemappings(remappings, imported);
 }
 
 /**
- * Given a partial map `files` from file names to file contents, a list of
- * `remappings` and a list of `ImportResolver`s `resolvers`, find and return all
- * ADDITIONAL files that are imported from the starting set `files` but are
- * missing in `files`. The return value is also a map from file names to file
- * contents.
+ * Given a partial map `files` from **source unit names** to file contents, a list of
+ * `remappings` and a list of `ImportResolver`s - `resolvers`, find all
+ * files that are imported from the starting set `files` but are
+ * **missing** in `files` and add them into the files map.
  */
 export function findAllFiles(
     files: Map<string, string>,
     remappings: Remapping[],
     resolvers: ImportResolver[]
-): Map<string, string> {
-    const queue: Array<[string | undefined, string]> = [];
+): void {
+    // Queue of source unit names to process
+    const queue: string[] = [];
 
     for (const fileName of files.keys()) {
-        queue.push([undefined, fileName]);
+        queue.push(fileName);
     }
```

Comment:
Wondering if spread will work here:
```typescript
const queue = [...files.keys()];
```

Reply:
Fixed in [eda0998](https://github.com/ConsenSys/solc-typed-ast/pull/93/commits/eda09983d90f42073d7083baf8f79bd68802792f)